controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java

```diff
@@ -683,14 +753,80 @@ public static boolean isScaleInputValid(final List<Integer> segmentsToSeal,
+        }
+
+        return epochStreamCutMap;
+    }
+
+    private static Set<Integer> computeToDelete(Map<Segment, Integer> epochCutMap, byte[] historyTable,
+                                                byte[] segmentTable, Set<Integer> deletedSegments) {
+        Set<Integer> toDelete = new HashSet<>();
+        int highestEpoch = epochCutMap.values().stream().max(Comparator.naturalOrder()).orElse(Integer.MIN_VALUE);
+
+        Optional<HistoryRecord> historyRecordOpt = HistoryRecord.readRecord(historyTable, 0, true);
+
+        // start with epoch 0 and go all the way upto epochCutMap.highEpoch
+        while (historyRecordOpt.isPresent() && historyRecordOpt.get().getEpoch() <= highestEpoch) {
+            HistoryRecord historyRecord = historyRecordOpt.get();
+            int epoch = historyRecord.getEpoch();
+
+            toDelete.addAll(historyRecord.getSegments().stream().filter(epochSegmentNumber -> {
+                Segment epochSegment = getSegment(epochSegmentNumber, segmentTable);
+                // ignore already deleted segments from todelete
+                // toDelete.add(epoch.segment overlaps cut.segment && epoch < cut.segment.epoch)
+                return !deletedSegments.contains(epochSegmentNumber) &&
+                        epochCutMap.entrySet().stream().noneMatch(cutSegment -> cutSegment.getKey().getNumber() == epochSegment.getNumber() ||
+                        (cutSegment.getKey().overlaps(epochSegment) && cutSegment.getValue() <= epoch));
+            }).collect(Collectors.toSet()));
+            historyRecordOpt = HistoryRecord.fetchNext(historyRecord, historyTable, true);
+        }
+        return toDelete;
+    }
+
+    private static boolean greaterThan(Map<Segment, Integer> map1, Map<Segment, Integer> map2, Map<Integer, Long> cut1, Map<Integer, Long> cut2) {
+        // find overlapping segments in map2 for all segments in map1
+        // compare epochs. map1 should have epochs gt or eq its overlapping segments in map2
+        return map1.entrySet().stream().allMatch(e1 ->
+                map2.entrySet().stream().noneMatch(e2 ->
+                        (e2.getKey().getNumber() == e1.getKey().getNumber() && cut1.get(e1.getKey().getNumber()) < cut2.get(e2.getKey().getNumber()))
+                        || (e2.getKey().overlaps(e1.getKey()) && e1.getValue() < e2.getValue())));
+    }
+
     /**
      * Helper method to compute list of continuous ranges. For example, two neighbouring key ranges where,
      * range1.high == range2.low then they are considered neighbours.
      * This method reduces input range into distinct continuous blocks.
      * @param input list of key ranges.
      * @return reduced list of key ranges.
      */
-    private static List<AbstractMap.SimpleEntry<Double, Double>> reduce(List<AbstractMap.SimpleEntry<Double, Double>> input) {
+    public static List<AbstractMap.SimpleEntry<Double, Double>> reduce(List<AbstractMap.SimpleEntry<Double, Double>> input) {
```

Comment:
I made this public because we use this to validate a stream cut. If it is legitimate or not.
Let me make a new public method for it which internally uses this method.